app/autosynccontroller.cpp

```diff
@@ -8,33 +8,107 @@
  ***************************************************************************/
 
 #include "autosynccontroller.h"
+#include "coreutils.h"
 
-AutosyncController::AutosyncController( QObject *parent ) : QObject( parent )
-{
-  // raise NotImplemented :)
-}
+#include "qgsproject.h"
+#include "qgsvectorlayer.h"
 
-AutosyncController::~AutosyncController()
+AutosyncController::AutosyncController(
+  LocalProject *openedProject,
+  QgsProject *openedQgsProject,
+  QObject *parent
+)
+  : QObject( parent )
+  , mActiveProject( openedProject )
+  , mActiveQgsProject( openedQgsProject )
 {
-  // raise NotImplemented :)
+  if ( !mActiveQgsProject || !mActiveProject )
+  {
+    CoreUtils::log( QStringLiteral( "Autosync" ), QStringLiteral( "Received an invalid active project data" ) );
+    return;
+  }
+
+  // Register for data change of project's vector layers
+  for ( const QgsMapLayer *layer : mActiveQgsProject->mapLayers( true ) )
+  {
+    const QgsVectorLayer *vecLayer = qobject_cast<const QgsVectorLayer *>( layer );
+    if ( vecLayer )
+    {
+      QObject::connect( vecLayer, &QgsVectorLayer::afterCommitChanges, this, &AutosyncController::handleLocalChange );
+    }
+  }
+
+  if ( mActiveProject->localVersion < 0 )
+  {
+    // this is not a mergin project
+    setSyncStatus( SyncStatus::NotAMerginProject );
+  }
 }
 
+AutosyncController::~AutosyncController() = default;
+
 AutosyncController::SyncStatus AutosyncController::syncStatus()
 {
   return mSyncStatus;
 }
 
-void AutosyncController::setActiveProject( Project )
+void AutosyncController::synchronizationProgressed( const QString &projectFullName, qreal progress )
+{
+  if ( !mActiveProject )
+    return;
+
+  QString activeProjectName = MerginApi::getFullProjectName( mActiveProject->projectNamespace, mActiveProject->projectName );
+
+  if ( projectFullName == activeProjectName )
+  {
+    if ( progress >= 0 && progress < 1 )
+    {
+      setSyncStatus( SyncStatus::SyncInProgress );
+    }
+  }
+}
+
+void AutosyncController::synchronizationFinished( const QString &, const QString &projectFullName, bool successfully, int )
 {
-  // raise NotImplemented :)
+  if ( !mActiveProject )
+    return;
+
+  QString activeProjectName = MerginApi::getFullProjectName( mActiveProject->projectNamespace, mActiveProject->projectName );
```

Comment:
I do not understand these checks all over the code. How it can happen that projectFullName != activeProjectName if AutoSync controller is destroyed on project change?